[flang][RFC] Adding a design document for assumed-rank objects

jeanPerier opened the pull request:
This patch adds a document describing assumed-rank objects and the related features as well as how they will be implemented in Flang.

jeanPerier committed:
[flang][RFC] Adding a design document for assumed-rank objects

This patch adds a document describing assumed-rank objects and the
related features as well as how they will be implemented in Flang.

klausler commented on flang/docs/AssumedRank.md:
```diff
@@ -0,0 +1,632 @@
+-->
+# Assumed-Rank Objects
+
+An assumed-rank is a data object that takes its rank from its effective
```
`An assumed rank dummy data object is a dummy argument that ...`

klausler commented on flang/docs/AssumedRank.md:
```diff
@@ -0,0 +1,632 @@
+# Assumed-Rank Objects
+
+An assumed-rank is a data object that takes its rank from its effective
+argument. It is a dummy, or the associated entity of a SELECT RANK in the `RANK
```
`It is a dummy argument, or ...`

klausler commented on flang/docs/AssumedRank.md:
```diff
@@ -0,0 +1,632 @@
+
+An assumed-rank is a data object that takes its rank from its effective
+argument. It is a dummy, or the associated entity of a SELECT RANK in the `RANK
+DEFAULT` or `RANK(*)` blocks. Its rank is not known at compile time. The rank
```
Just `RANK DEFAULT`.  In a `RANK(*)` block, the construct entity is not assumed-rank, but rather an assumed-size vector with bounds `(1:*)`.

jeanPerier replied:
Thanks, I read the RANK(*) case too fast. I updated the descriptions and example.

Note that my understanding is that the descriptor created when passing an assumed-size actual to an assumed-rank dummy should still have the assumed-size rank (not forced to 1 at that point), and that it is only when reaching a RANK(*) that the original rank of the assumed-size should be "lost".

jeanPerier replied:
> I seems best to change the rank to 1 at the first call;

I agree but is it legal?

I think this would prevent `RANK(x)` to return the rank of the effective argument of the assumed rank as requested by the standard (there is no special case for assumed-rank associated with assumed-size in RANK(), nor in 8.5.8.7).
UBOUND and al. also does not say that an assumed-rank associated to an assumed-size has rank one (only mention a special case for the "final dimension").

I think we can only change it in the RANK(*), even if it makes all SELECT RANK checking a bit more expensive (especially since we must not fall in RANK(cst) for an assumed-size even if there is not RANK(*)).

All other compilers that can compile the program below seem to keep the rank of the assumed-size attached to the assumed-rank:

````
module m
contains
subroutine print_my_rank(x)
  real :: x(..)
  print *, rank(x)
end subroutine

subroutine test(x1, x2, x3)
 real :: x1(*)
 real :: x2(1, *)
 real :: x3(1, 1, *)
 call print_my_rank(x1)
 call print_my_rank(x2)
 call print_my_rank(x3)
end subroutine
end module

 use m
 real :: x1(1)
 real :: x2(1, 1)
 real :: x3(1, 1, 1)
 call test(x1, x2, x3)
 end
```

Prints "1,2,3".

klausler commented on flang/docs/AssumedRank.md:
```diff
@@ -0,0 +1,632 @@
+represent assumed-rank in procedure characteristics.
+
+### Runtime Representation of Assumed-Ranks
+Assumed-ranks are implemented as CFI_cdesct_t (18.5.3) with the addition of an
```
`CFI_cdesc_t`

klausler commented on flang/docs/AssumedRank.md:
```diff
@@ -0,0 +1,632 @@
+for instance) while the stack "over" allocation is likely reasonable (24 bytes
+per dimension). Hence the selection of the simple approach using static size
+allocation to the maximum rank.
+
```
When an assumed-size dummy argument is associated with an assumed-rank dummy argument, how will the callee use the descriptor to distinguish `RANK(1)` from `RANK(*)` in a `SELECT RANK` construct?  Maybe because the descriptor looks like a rank-1 vector with bounds of `(1:-1)`?  Note that `UBOUND(assumedrank,1)` is defined to be -1 when the assumed-rank dummy argument is associated with an assumed-size effective argument (F'2023 16.9.215).

EDIT: Answered below, never mind.

klausler commented on flang/docs/AssumedRank.md:
```diff
@@ -0,0 +1,632 @@
+involved and describes the related implementation design.
+
+### Assumed-rank in procedure references
+Assumed-rank arguments are implemented as being the address of a CFI_cdesct_t.
```
`CFI_cdesc_t`

klausler commented on flang/docs/AssumedRank.md:
```diff
@@ -0,0 +1,632 @@
+- Copy-in/copy-out when required
+- Creation of forwarding of the assumed-rank dummy descriptor
+- Finalization, deallocation, and initialization of INTENT(OUT) assumed-rank
+  dummy.
```
Also: passing an assumed-size dummy array to an assumed-rank dummy data object.

klausler commented on flang/docs/AssumedRank.md:
```diff
@@ -0,0 +1,632 @@
+
+#### Intent(out) assumed-rank finalization, deallocation, initialization
+
+C839 is limiting the cases where assumed-rank may be finalizable, but I do not
```
F'2023 7.5.6.3 p7: only non-allocatable non-pointer `INTENT(OUT)` dummy arguments are finalized.  8.5.10 p3 states that pointer dummy arguments with `INTENT(OUT)` just have their association status set to disassociated (i.e., nullified, not deallocated).

jeanPerier replied:
Right, but C839 talks about "_nonallocatable nonpointer assumed-rank array is an **actual** argument_".

I thought if was a typo in the standard where the "nonallocatable nonpointer" part should have applied to the dummy later in the sentence, but thinking this through a bit more, I think the restriction on nonallocatable nonpointer assumed-rank actuals comes from the fact that they _may_ be associated with an assumed-size array which would make any finalization attempt of the dummy a bug.

Interesting though that the standard did not push this corner case to the user as a constraint like "actual assumed-rank must not be associated to an assumed-size".

I updated the section with this explanation instead of my own ignorance.

klausler commented on flang/docs/AssumedRank.md:
```diff
@@ -0,0 +1,632 @@
+detects assumed-ranks associated with an assumed-size arrays regardless of the
+rank, and takes precedence over any rank based matching.
+
+Note that `-1` is a magic extent number that encodes that a descriptor describe
```
This answers my question above.

klausler reviewed (approved):
LGTM, ship it.

jeanPerier committed:
address review comments

- update RANK(*) description and example with the understanding that
  the related associate entity has rank one (instead of being an
  assumed-rank as previously described).
- update C839 constraint description
- various typo corrections, and double spaces elimination.

kkwli commented on flang/docs/AssumedRank.md:
```diff
@@ -0,0 +1,648 @@
+maximized using the maximum rank, which the runtime code already does when
+creating temporary descriptor in many cases. Inline code also needs care if it
+needs to access the descriptor addendum (like the type descriptor), since its
+offset will not be a compile time constant as usual.
```
Would it also affect privatization in OpenMP or OpenACC? We always create a private descriptor with maximum rank. 
```fortran
subroutine f(x)
  real :: x(..)
!$omp parallel private(x)
  select rank (z => x)
  rank(2)
    x(i,j) = ...
  rank(3)
    x(i,j,k) = ...
  end select
!$omp end parallel
end
```
How about the implication for mapping?

jeanPerier replied:
You will indeed need to create a private descriptor with maximum rank and pay attention to not blindly copy the whole incoming descriptor as if it has the maximum rank (the incoming descriptor storage may be smaller since it was allocated with the actual rank knowledge).

Is this happening in [createHostAssociateVarClone ](https://github.com/llvm/llvm-project/blob/5107a47fee9c95489d3743a53e9578407b0e3656/flang/lib/Lower/Bridge.cpp#L610C8-L610C35)inside Bridge.cpp, or is this private copy made later in some OpenMP IR pass?

kiranchandramohan replied:
An advise here on the implementation will be useful.

jeanPerier replied:
> An advise here on the implementation will be useful.

It depends where this is implemented: is this abstracted in the MLIR and only made explicit in when generating LLVM IR, or is the copy already created in lowering in createHostAssociateVarClone?

If it is in createHostAssociateVarClone, the burden of the copy complexity will be borne by the fir.load/fir.store operations. So you can do an copy of the descriptor in the IR as:
```
  %new_box_addr = fir.alloca !fir.box<fir.array<..xT>> ! allcoates maximum ranked descriptor storage
  %fir.store %old_box to %new_box_addr ! does a "non trivial" dynamic copy
```

kiranchandramohan reviewed (commented):
Thanks for this RFC. I have a few minor points/questions.

-> Does this have an effect on the processing of the ignore_tkr directive?
-> I think the new mpi module (mpi_f08) uses assumed rank arrays. Some early references talk about assumed-rank as a request from MPI (https://j3-fortran.org/doc/year/09/09-100.txt, https://j3-fortran.org/doc/year/08/08-271.txt). Might be reasonable test for the feature.
-> Please add the document to the index file (https://github.com/llvm/llvm-project/blob/main/flang/docs/index.md).

jeanPerier committed:
add mention of IGNORE_TKR and MPI_F08. Remove fir tags since they have no lexer

jeanPerier commented:
> Does this have an effect on the processing of the ignore_tkr directive?

Good question. No more than when passing to an assumed-shape.
- It is forbidden (and pointless) to use IGNORE_TKR(r) on an assumed rank dummy
- It is possible to use IGNORE_TKR(tk)  on an assumed rank dummy (but highly preferable to use TYPE(*)). In which case on caller side, the dynamic type information in the descriptor will be set according to the actual. But just like with assumed shapes, if the procedure with ignore_tkr is implemented in Fortran, it is quite unspecified what type it will be considered to have inside the procedure. Any direct usages of the incoming descriptor in the Fortran runtime will behave according to the actual argument type, while any new inline and new descriptor created will use the dummy type information.
- It is possible to pass an assumed-rank object to an explicit shape/assumed-size array dummy that has IGNORE_TKR(R). In which case, lowering will do copy_in/copy_out as when passing to a contiguous assumed-rank, and then pass only the base address instead of the descriptor.

>  I think the new mpi module (mpi_f08) uses assumed rank arrays.

Indeed, there is mention of assumed rank in the MPI standard https://www.mpi-forum.org/docs/mpi-3.1/mpi31-report.pdf
I added a note in annex about using mpi_f08 to test for this feature, thanks for pointing to it.

> Please add the document to the index file (https://github.com/llvm/llvm-project/blob/main/flang/docs/index.md).

Done, thanks.

jeanPerier commented on flang/docs/AssumedRank.md:
````diff
@@ -0,0 +1,648 @@
+
+Pseudo FIR for the example (some converts and SSA constants creation are not shown for more clarity):
+
+```FIR
````
Done.

kiranchandramohan reviewed (approved):
LGTM. Thanks.
